src/dev/precommit_hook/check_file_casing.js

```diff
@@ -0,0 +1,54 @@
+import { createFailError } from '../run';
+
+const NON_SNAKE_CASE_RE = /[A-Z \-]/;
+const ALLOW_NON_SNAKE_CASE = [
+  'docs/**/*',
+  'packages/eslint-*/**/*',
+  'src/babel-*/**/*',
+  'tasks/config/**/*',
+];
+
+function listFileNames(files) {
+  return files
+    .map(file => ` - ${file.getRelativePath()}`)
+    .join('\n');
+}
+
+/**
+ * Check that all passed File objects are using valid casing. Every
+ * file SHOULD be using snake_case but some files are allowed to stray:
+ *
+ *  - grunt config: the file name needs to match the module name
+ *  - eslint/babel packages: the directory name matches the module
+ *    name, which uses kebab-case to mimic other babel/eslint plugins,
+ *    configs, and presets
+ *  - docs: unsure why, but all docs use kebab-case and that's fine
+ *
+ * @param {ToolingLog} log
+ * @param {Array<File>} files
+ * @return {Promise<undefined>}
+ */
+export async function checkFileCasing(log, files) {
+  const errors = [];
+  const warnings = [];
+
+  files.forEach(file => {
+    const invalid = file.matchesRegex(NON_SNAKE_CASE_RE);
+    const allowed = file.matchesAnyGlob(ALLOW_NON_SNAKE_CASE);
+    if (!invalid) {
+      log.debug('%j uses valid casing', file);
+    } else if (allowed) {
+      warnings.push(file);
+    } else {
+      errors.push(file);
+    }
+  });
+
+  if (warnings.length) {
```

Comment:
Warnings won't cause a failure, which is why they are "should be", and errors do, which is why they are "must be". I'm adding the warnings because I think it's easy to create files that are not using snake_case without realizing it (in situations where they are technically allowed, but not *necessary*) and hope that the warning will draw enough attention to encourage a correction.

Reply:
Sorry it was a silly English grammar question :) I was just wondering why in one case `filenames modal_verb *be* snake_case`, but in other case `filenames modal_verb *use* snake_case`